File: 2020/08/Main.hs
import qualified Data.Vector                   as V
import           Data.Vector                    ( Vector )
import           System.Environment
import           Text.Printf

getOperation :: String -> String
getOperation = head . words

getArgument :: String -> Int
getArgument = read . dropWhile (== '+') . last . words

stepMachine :: [Int] -> Int -> Int -> Vector String -> (Int, Bool)
stepMachine y i a x | i == length x = (a, True)
                    | i `elem` y    = (a, False)
                    | otherwise     = stepMachine y' i' a' x
 where
  instruction = x V.! i
  operation   = getOperation instruction
  argument    = getArgument instruction
  y'          = i : y
  i' | operation == "jmp" = i + argument
     | otherwise          = i + 1
  a' | operation == "acc" = a + argument
     | otherwise          = a

getFixIndices :: Vector String -> [Int]
getFixIndices =
  V.toList
    . V.map fst
    . V.filter ((`elem` ["jmp", "nop"]) . getOperation . snd)
    . V.indexed

swapOperation :: Vector String -> Int -> Vector String
swapOperation v i = v'
 where
  this      = v V.! i
  operation = getOperation this
  argument  = last . words $ this
  next | operation == "jmp" = "nop " ++ argument
       | otherwise          = "jmp " ++ argument
  v' = v V.// [(i, next)]

findFixed :: [Vector String] -> Int
findFixed [] = 0
findFixed (x : xs) | fixed     = acc
                   | otherwise = findFixed xs
  where (acc, fixed) = stepMachine [] 0 0 x

part1 :: [String] -> Int
part1 = fst . stepMachine [] 0 0 . V.fromList

part2 :: [String] -> Int
part2 x = findFixed . map (swapOperation vector) . getFixIndices $ vector
  where vector = V.fromList x

output :: String -> IO ()
output path = do
  input <- filter (not . null) . lines <$> readFile path
  printf "File: %s\n" path
  printf "  Part 1: %d\n" . part1 $ input
  printf "  Part 2: %d\n" . part2 $ input

main :: IO ()
main = getArgs >>= mapM_ output

File: 2020/08/Rust/src/main.rs
use std::convert::TryFrom;
use std::env;
use std::fs;

fn parse_input(input: &str) -> Vec<&str> {
    input.lines().filter(|x| !x.is_empty()).collect()
}

fn get_instruction(instruction: &str) -> (&str, isize) {
    let instruction = instruction.split_whitespace().collect::<Vec<_>>();
    let argument = instruction[1].parse().unwrap();
    (instruction[0], argument)
}

fn step_machine(instructions: &[&str]) -> (isize, bool) {
    let mut executed = Vec::new();
    let mut index: isize = 0;
    let mut acc: isize = 0;
    let convert_num = |x| usize::try_from(x).unwrap();
    while convert_num(index) != instructions.len() {
        if executed.contains(&index) {
            return (acc, false);
        }
        let instruction = instructions[convert_num(index)];
        let (operation, argument) = get_instruction(instruction);
        executed.push(index);
        if operation == "jmp" {
            index += argument;
        } else if operation == "acc" {
            acc += argument;
            index += 1;
        } else {
            index += 1;
        }
    }
    (acc, true)
}

fn get_fix_indices(instructions: &[&str]) -> Vec<usize> {
    instructions
        .iter()
        .enumerate()
        .filter(|x| {
            let (operation, _) = get_instruction(x.1);
            if ["jmp", "nop"].contains(&operation) {
                return true;
            }
            false
        })
        .map(|x| x.0)
        .collect::<Vec<_>>()
}

fn swap_operation(instructions: &[String], index: usize) -> Vec<String> {
    let this = instructions[index].clone();
    let (operation, argument) = get_instruction(&this);
    let concat_num = |x: &str, y: isize| -> String {
        let mut x = x.to_string();
        x.push_str(&y.to_string());
        x
    };
    let next = {
        if operation == "jmp" {
            concat_num("nop ", argument)
        } else {
            concat_num("jmp ", argument)
        }
    };
    let mut instructions = instructions.to_owned();
    instructions[index] = next;
    instructions.iter().map(|x| x.to_string()).collect()
}

fn find_fixed(set: &[String]) -> isize {
    let mut temp = Vec::new();
    for t in set {
        temp.push(&t[..]);
    }
    let (acc, fixed) = step_machine(&temp);
    if fixed {
        return acc;
    }
    0
}

fn part_2(instructions: &[&str]) -> isize {
    let indices = get_fix_indices(instructions);
    let instructions = instructions
        .iter()
        .map(|x| x.to_string())
        .collect::<Vec<_>>();
    for index in indices {
        let acc = find_fixed(&swap_operation(&instructions, index));
        if acc != 0 {
            return acc;
        }
    }
    0
}

fn output(path: &str) {
    let input = fs::read_to_string(&path).unwrap();
    let parsed = parse_input(&input);

    println!("File: {}", path);
    println!("  Part 1: {}", step_machine(&parsed).0);
    println!("  Part 2: {}", part_2(&parsed));
}

fn main() {
    env::args()
        .skip(1)
        .collect::<Vec<_>>()
        .iter()
        .for_each(|x| output(x));
}
